File: java/com/joshmanisdabomb/lcc/block/network/CogNetwork.java
package com.joshmanisdabomb.lcc.block.network;

import com.joshmanisdabomb.lcc.block.CogBlock;
import com.joshmanisdabomb.lcc.misc.Util;
import net.minecraft.block.BlockState;
import net.minecraft.block.RedstoneWireBlock;
import net.minecraft.state.properties.BlockStateProperties;
import net.minecraft.util.Direction;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.apache.commons.lang3.tuple.ImmutablePair;
import org.apache.commons.lang3.tuple.Pair;

import java.util.ArrayList;
import java.util.List;

public class CogNetwork extends BlockNetwork<Pair<BlockPos, Direction>> {

    public CogNetwork(int distance) {
        super(distance);
    }

    @Override
    protected List<Pair<BlockPos, Direction>> traverse(World world, Pair<BlockPos, Direction> current, List<BlockPos> nodes) {
        ArrayList<Pair<BlockPos, Direction>> positions = new ArrayList<>();
        BlockPos pos = this.toPosition(current);
        BlockState state = world.getBlockState(pos);
        Direction d = current.getRight();
        if (d == null) {
            ((CogBlock)state.getBlock()).getCogs(state).forEach(cd -> positions.add(new ImmutablePair<>(pos, cd)));
        } else {
            for (Direction pd : Util.PERPENDICULARS.get(d)) {
                BlockPos adjacentPos = pos.offset(pd);
                BlockState adjacent = world.getBlockState(adjacentPos);
                BlockPos wrapPos = adjacentPos.offset(d);
                BlockState wrap = world.getBlockState(wrapPos);

                if (((CogBlock)state.getBlock()).hasCog(pd, state)) positions.add(new ImmutablePair<>(pos, pd));

                if (adjacent.getBlock() instanceof CogBlock) {
                    if (((CogBlock)adjacent.getBlock()).hasCog(d, adjacent)) positions.add(new ImmutablePair<>(adjacentPos, d));
                } else if (adjacent.canProvidePower() && (world.getRedstonePower(adjacentPos, pd) > 0 || (adjacent.getBlock() instanceof RedstoneWireBlock && adjacent.get(BlockStateProperties.POWER_0_15) > 0))) {
                    nodes.add(adjacentPos);
                }

                if (!adjacent.isNormalCube(world, adjacentPos) && wrap.getBlock() instanceof CogBlock && ((CogBlock)wrap.getBlock()).hasCog(pd.getOpposite(), wrap)) positions.add(new ImmutablePair<>(wrapPos, pd.getOpposite()));
            }
        }
        return positions;
    }

    @Override
    protected BlockPos toPosition(Pair<BlockPos, Direction> traversable) {
        return traversable.getLeft();
    }

}

File: lcc-content/src/main/kotlin/com/joshmanisdabomb/lcc/network/CogNetwork.kt
package com.joshmanisdabomb.lcc.network

import com.joshmanisdabomb.lcc.block.CogBlock
import com.joshmanisdabomb.lcc.block.CogBlock.Companion.cog_states
import com.joshmanisdabomb.lcc.directory.LCCBlocks
import com.joshmanisdabomb.lcc.extensions.perpendiculars
import net.minecraft.block.SideShapeType
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.BlockView

class CogNetwork(distance: Int = 64) : BlockNetwork<Pair<BlockPos, Direction?>>(distance) {

    override fun traverse(world: BlockView, current: Pair<BlockPos, Direction?>, nodes: MutableMap<String, MutableSet<Pair<BlockPos, Direction?>>>): Set<Pair<BlockPos, Direction?>> {
        val pos = toPosition(current)
        val state = world.getBlockState(pos)
        val d = current.second
        if (state.block !is CogBlock) return emptySet()
        if (d == null) return directions.mapNotNull { if (state.get(cog_states[it]).exists) pos to it else null }.toSet()

        val me = pos to d
        var toClockwise = false
        if (from(nodes, "ccw").contains(me)) toClockwise = true

        //when powered from behind, add node in network result
        if (LCCBlocks.cog.isPowered(state, world, pos, d) == true) {
            from(nodes, "powered").add(me)
            //from(nodes, "ccw").remove(me)
        }

        //check connected cogs
        val positions = mutableSetOf<Pair<BlockPos, Direction?>>()
        for (d2 in d.perpendiculars) {
            if (state[cog_states[d2]].exists) {
                val pair = pos to d2
                positions.add(pair)
                if (!toClockwise) from(nodes, "ccw").add(pair)
            }

            val p1 = pos.offset(d2)
            val c1 = world.getBlockState(p1)
            if (c1.block is CogBlock && c1[cog_states[d]].exists) {
                val pair = p1 to d
                positions.add(pair)
                if (!toClockwise) from(nodes, "ccw").add(pair)
            }

            if (!c1.isSideSolid(world, pos.offset(d2), d.opposite, SideShapeType.FULL)) {
                val p2 = pos.offset(d2).offset(d)
                val c2 = world.getBlockState(p2)
                if (c2.block is CogBlock && c2[cog_states[d2.opposite]].exists) {
                    val pair = p2 to d2.opposite
                    positions.add(pair)
                    if (!toClockwise) from(nodes, "ccw").add(pair)
                }
            }
        }
        return positions
    }

    override fun toPosition(traversable: Pair<BlockPos, Direction?>): BlockPos = traversable.first

}